File: leet_code/go/countingBits.go
import "math"

func countBits(num int) []int {
	currentPower := 1
	var ans []int
	ans = append(ans, 0)
	for i := 1; i <= num; i++ {
		if i > int(math.Pow(float64(2), float64(currentPower))) {
			ans = append(ans, ans[len(ans) - 1] + 1)
			currentPower++
		} else if i == int(math.Pow(float64(2), float64(currentPower))) {
			ans = append(ans, 1)
		} else if i == int(math.Pow(float64(2), float64(currentPower))) - 1 {
			ans = append(ans, currentPower)
		} else {
			idx := i - int(math.Pow(float64(2), float64(currentPower - 1)))
			ans = append(ans, ans[idx] + ans[int(math.Pow(float64(2), float64(currentPower - 1)))])
		}
	}
	return ans
}

File: leet_code/ruby/counting_bits.rb
# @param {Integer} num
# @return {Integer[]}
def count_bits(num)
	current_power = 1
	ans = [0]
	(1..num).each do |n|
		if n > 2**current_power
			ans.push(ans.last + 1)
			current_power += 1
		elsif n == 2**current_power
			ans.push(1)
		elsif n == 2**current_power - 1
			ans.push(current_power)
		else
			idx = n - 2**(current_power - 1)
			ans.push(ans[idx] + ans[2**(current_power - 1)])
		end

	end 
	ans
end